how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py

```diff
@@ -85,17 +95,43 @@ def train(net, trainloader, run, rank, params):
 
 def test(net, testloader, run, rank):
 
+    classes = [
+        "airplane",
+        "automobile",
+        "bird",
+        "cat",
+        "deer",
+        "dog",
+        "frog",
+        "horse",
+        "ship",
+        "truck",
+    ]
+
     correct = 0
     total = 0
 
     with torch.no_grad():
-        for data in testloader:
+        for i, data in enumerate(testloader):
             images, labels = data
             images, labels = images.to(f"cuda:{rank}"), labels.to(f"cuda:{rank}")
             outputs = net(images)
             _, predicted = torch.max(outputs.data, 1)
             total += labels.size(0)
 
+            if i % 10 == 0:
+                probs = F.softmax(outputs, dim=1)
+                probs = probs.cpu().numpy()
+
+                for i, ps in enumerate(probs):
+                    pred = classes[np.argmax(ps)]
+                    gt = classes[labels[i]]
+                    # (Neptune) Save images and predictions from every node
```

Comment:
Is this the crucial code block that was mentioned?

Reply:
Yap, but the emphasis is on the run. We log metadata from each node/process using `rank` as the namespace separator. It avoids concurrency and overriding issues.